Write Biopython SeqRecords as colourful word docs

DAWells opened the pull request:
The new function `write_crazydoc()` will save one or more SeqRecords to a word doc with features formatted with a combination of highlight colours, font colours, bold, underline, and italics.

The type of formatting applied is based on feature overlap because overlapping features cannot use the same type of formatting (a single base/residue cannot have more than one font colour). This means that there is a limit to feature annotation complexity which can be converted to word doc. An informative assertion error will be thrown if each feature cannot be correctly formatted, this is verified in the new test file. Custom colour combinations can be specified to overcome this as demonstrated in the `write_crazydoc` doc string. The helper function `_test_colours()` can be used to check colour combinations as potential highlight-font colour combinations are hard to read.

There is also a test added which verifies the locations of features are not changed by writing and re-reading a crazydoc.

veghp committed:
Merge pull request #2 from Edinburgh-Genome-Foundry/dev

Biopython fix + protein sequence support

veghp committed:
Merge branch 'dev'

DAWells committed:
Write SeqRecords to docx with features formatted

_features_per_position list each feature overlapping each position.
make_groups() assigns features to 5 groups so
that overlapping features are in different groups.
_make_feature_format assigns compatible formats to each feature.
write_crazydoc() Saves a list of sequence to a docx with features formatted.

DAWells committed:
Fix #3 Informative errors if features not formatted

DAWells committed:
Allow supplied formats and test colours. close #1

DAWells committed:
Improve _check_all_features_grouped error message

DAWells committed:
Reorder highlight colours

DAWells committed:
Add example write_crazydoc to readme

Fix #4

DAWells committed:
write_crazydoc with custom formats.
write_position_features forces upper case

DAWells committed:
Add tests for crazydoc writer

- Confirms that reading,writing,reading a crazydoc
doesn't change feature locations
- Confirms write_crazydoc will not run if
too many groups, or too many features to format correctly

DAWells committed:
Update README

Correct crazydoc writer command and add import
fix #4

DAWells committed:
Re-order highlight colours

DAWells committed:
Move red highlight lower in order

DAWells committed:
Merge pull request #5 from DAWells/gb2doc

Gb2doc
Allow writing biopython seqrecords to word doc with formatting to separate features

This can handle a certain amount of feature overlap but is limited by the number of formatting colours. Will raise an exception if there are too many overlaps to colour distinctly.

veghp commented:
Thank you for this feature addition. I'll have to review and test the code. I have some reservations about converting sequences into a more limited format, but I think it may be useful for creating documents where we don't want to insert images.

Zulko commented:
This is fun! Out of curiosity, did you _really need_ a way to generate docs? If so I'd like to know the back story! My main comments are
- to run pep8 or Black on the code, in particular to keep lines below 88-92 characters for style coherence,
- the filename `CrazyDocWriters.py` lets us think that there is a class in there called `CrazyDocWriter`. You could actually write the code as a class, but the shortest path here would probably be to rename the file as `crazydoc_writers.py`.

DAWells committed:
rename crazydoc_writers module

pep8 naming convension is lower case for modules,
classes use capitalised words

DAWells committed:
Update crazydoc_writers module name in __init__.py

DAWells committed:
Use `black` to comply with pep8

DAWells commented:
Thanks for the feedback :smile: I've now:
- used Black to follow pep8 style
- renamed `CrazydocWriters.py` to `crazydoc_writers.py`

My motivation was just having to share annotated sequences with people who want them in word docs. This seemed like a faster and more accurate solution than doing in manually every time. Plus figuring out how to format overlapping sequences seemed like an interesting problem!
